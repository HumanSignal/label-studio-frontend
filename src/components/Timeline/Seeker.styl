.seeker
  width 100%
  height 24px
  position relative

  &__track
    width 100%
    height 100%
    background #FAFAFA
    box-shadow inset 0px 1px 0px rgba(0, 0, 0, 0.1), inset 0px 0px 0px 1px rgba(0, 0, 0, 0.15)
    border-radius 3px

  &__indicator
    top 0
    left 0
    z-index 1
    width 100px
    height 24px
    max-width 100%
    position absolute
    border-radius 3px
    cursor col-resize
    box-shadow 0 0 0 1px #000, 0 0 0 1px #000 inset

  &__position
    top 0
    left 0
    width 1px
    z-index 2
    cursor grab
    height 100%
    position absolute
    background-color #000

    &::before
      top 0
      bottom 0
      left -5px
      right -5px
      content ''
      display block
      position absolute

    &:hover
      border-radius 30px
      box-shadow 0 0 0 3px rgba(#000, 0.2)

  &__minimap
    top 0
    left 0
    z-index 0
    width 100%
    height 100%
    position absolute
    pointer-events none
