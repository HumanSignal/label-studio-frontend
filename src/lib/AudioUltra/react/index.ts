import { MutableRefObject, useEffect, useRef, useState } from "react";
import { Waveform, WaveformOptions } from "../Waveform";

export const useWaveform = (
  containter: MutableRefObject<HTMLElement | null | undefined>,
  options: Omit<WaveformOptions, "container"> & { onLoad?: (wf: Waveform) => void },
) => {
  const waveform = useRef<Waveform>();

  const [zoom, setZoom] = useState(1);
  const [volume, setVolume] = useState(1);
  const [playing, setPlaying] = useState(false);
  const [duration, setDuration] = useState(0);
  const [currentTime, setCurrentTime] = useState(0);
  const [amp, setAmp] = useState(1);
  const [rate, setRate] = useState(1);
  const [muted, setMuted] = useState(false);


  useEffect(() => {
    const wf = new Waveform({
      ...(options ?? {}),
      container: containter.current!,
    });

    waveform.current = wf;

    wf.load();

    wf.on("load", () => {
      setDuration(wf.duration);
      options?.onLoad?.(wf);
    });

    wf.on("play", () => setPlaying(true));
    wf.on("pause", () => setPlaying(false));
    wf.on("playing", setCurrentTime);
    wf.on("seek", setCurrentTime);
    wf.on("zoom", setZoom);
    wf.on("muted", setMuted);
    wf.on("volumeChange", setVolume);
    wf.on("rateChanged", setRate);

    return () => {
      wf.destroy();
    };
  }, []);

  useEffect(() => {
    const wf = waveform.current;

    if (wf && wf.loaded) {
      wf.zoom = zoom;
    }
  }, [zoom]);

  useEffect(() => {
    const wf = waveform.current;

    if (wf && wf.loaded) {
      wf.volume = volume;
    }
  }, [volume]);

  useEffect(() => {
    const wf = waveform.current;

    if (wf && wf.loaded) {
      wf.rate = rate;
    }
  }, [rate]);

  useEffect(() => {
    const wf = waveform.current;

    if (wf && wf.loaded) {
      wf.amp = amp;
    }
  }, [amp]);

  useEffect(() => {
    if (playing) {
      waveform.current?.play();
    } else {
      waveform.current?.pause();
    }
  }, [playing]);

  useEffect(() => {
    if (waveform.current) {
      waveform.current.muted = muted;
    }
  }, [muted]);

  return {
    waveform,
    zoom,
    setZoom,
    volume,
    setVolume,
    playing,
    setPlaying,
    duration,
    currentTime,
    setCurrentTime,
    amp,
    setAmp,
    rate,
    setRate,
    muted,
    setMuted,
  };
};
