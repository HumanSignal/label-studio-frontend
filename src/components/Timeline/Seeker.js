import { clamp } from "lodash";
import { useCallback } from "react";
import { Block, Elem } from "../../utils/bem";

import "./Seeker.styl";

export const Seeker = ({
  seek,
  length,
  seekOffset,
  seekVisible,
  onIndicatorMove,
  onSeek,
}) => {
  const width = `${seekVisible / length * 100}%`;
  const offsetLimit = length - seekVisible;
  const windowOffset = `${Math.min(seekOffset, offsetLimit) / length * 100}%`;
  const seekerOffset = seek / length * 100;

  const onIndicatorDrag = useCallback((e) => {
    const indicator = e.target;
    const startOffset = indicator.offsetLeft;
    const startDrag = e.pageX;
    const parentWidth = indicator.parentNode.clientWidth;

    const onMouseMove = (e) => {
      const limit = parentWidth - indicator.clientWidth;
      const newOffset = clamp(startOffset + (e.pageX - startDrag), 0, limit);
      const percent = newOffset / parentWidth;

      onIndicatorMove?.(Math.ceil(length * percent));
    };

    const onMouseUp = () => {
      document.removeEventListener("mousemove", onMouseMove);
      document.removeEventListener("mouseup", onMouseUp);
    };

    document.addEventListener("mousemove", onMouseMove);
    document.addEventListener("mouseup", onMouseUp);
  }, [length]);

  const onSeekerDrag = useCallback((e) => {
    const indicator = e.target;
    const startOffset = indicator.offsetLeft;
    const startDrag = e.pageX;
    const parentWidth = indicator.parentNode.clientWidth;

    const onMouseMove = (e) => {
      const limit = parentWidth - indicator.clientWidth;
      const newOffset = clamp(startOffset + (e.pageX - startDrag), 0, limit);
      const percent = newOffset / parentWidth;

      onSeek?.(Math.ceil(length * percent));
    };

    const onMouseUp = () => {
      document.removeEventListener("mousemove", onMouseMove);
      document.removeEventListener("mouseup", onMouseUp);
    };

    document.addEventListener("mousemove", onMouseMove);
    document.addEventListener("mouseup", onMouseUp);
  }, [length]);

  return (
    <Block name="seeker">
      <Elem name="track"/>
      <Elem name="indicator" style={{ left: windowOffset, width }} onMouseDown={onIndicatorDrag}/>
      <Elem name="position" style={{ left: `${seekerOffset}%` }} onMouseDown={onSeekerDrag}/>
    </Block>
  );
};
