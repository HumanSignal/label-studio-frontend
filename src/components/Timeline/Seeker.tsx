import { clamp } from "lodash";
import { FC, ReactElement, useCallback, useRef } from "react";
import { Block, Elem } from "../../utils/bem";
import { TimelineMinimapProps } from "./Types";

import "./Seeker.styl";

export interface SeekerProps {
  position: number;
  length: number;
  seekOffset: number;
  seekVisible: number;
  step: number;
  leftOffset?: number;
  minimap?: ReactElement<TimelineMinimapProps> | null;
  onIndicatorMove: (position: number) => void;
  onSeek: (position: number) => void;
}

export const Seeker: FC<SeekerProps> = ({
  position,
  length,
  seekOffset,
  seekVisible,
  onIndicatorMove,
  onSeek,
  minimap,
  step,
  ...props
}) => {
  const leftOffset = (props.leftOffset ?? 150) / step;
  const rootRef = useRef<HTMLDivElement>();
  const seekerRef = useRef<HTMLDivElement>();
  const viewRef = useRef<HTMLDivElement>();

  const showIndicator = seekVisible > 0;
  const width = `${(seekVisible - leftOffset) / length * 100}%`;
  const offsetLimit = length - (seekVisible - leftOffset);
  const windowOffset = `${Math.min(seekOffset, offsetLimit) / length * 100}%`;
  const seekerOffset = position / length * 100;

  const onIndicatorDrag = useCallback((e) => {
    const indicator = viewRef.current!;
    const dimensions = rootRef.current!.getBoundingClientRect();
    const indicatorWidth = indicator.clientWidth;

    const startDrag = e.pageX;
    const startOffset = startDrag - dimensions.left - (indicatorWidth / 2);
    const parentWidth = dimensions.width;
    const limit = parentWidth - indicatorWidth;

    const jump = clamp(Math.ceil(length * (startOffset / parentWidth)), 0, limit);

    onIndicatorMove?.(jump);

    const onMouseMove = (e: globalThis.MouseEvent) => {
      const newOffset = clamp(startOffset + (e.pageX - startDrag), 0, limit);
      const percent = newOffset / parentWidth;

      onIndicatorMove?.(Math.ceil(length * percent));
    };

    const onMouseUp = () => {
      document.removeEventListener("mousemove", onMouseMove);
      document.removeEventListener("mouseup", onMouseUp);
    };

    document.addEventListener("mousemove", onMouseMove);
    document.addEventListener("mouseup", onMouseUp);
  }, [length]);

  const onSeekerDrag = useCallback((e) => {
    const indicator = seekerRef.current!;
    const dimensions = rootRef.current!.getBoundingClientRect();
    const indicatorWidth = indicator.clientWidth;

    const startDrag = e.pageX;
    const startOffset = startDrag - dimensions.left - (indicatorWidth / 2);
    const parentWidth = dimensions.width;

    onSeek?.(clamp(Math.ceil(length * (startOffset / parentWidth)), 0, parentWidth));

    const onMouseMove = (e: globalThis.MouseEvent) => {
      const limit = parentWidth - indicator.clientWidth;
      const newOffset = clamp(startOffset + (e.pageX - startDrag), 0, limit);
      const percent = newOffset / parentWidth;

      onSeek?.(Math.ceil(length * percent));
    };
    const onMouseUp = () => {
      document.removeEventListener("mousemove", onMouseMove);
      document.removeEventListener("mouseup", onMouseUp);
    };

    document.addEventListener("mousemove", onMouseMove);
    document.addEventListener("mouseup", onMouseUp);
  }, [length]);

  const navigationHandler = showIndicator ? onIndicatorDrag : onSeekerDrag;

  return (
    <Block name="seeker" ref={rootRef} onMouseDown={navigationHandler}>
      <Elem name="track"/>
      {showIndicator && (
        <Elem name="indicator" ref={viewRef} style={{ left: windowOffset, width }}/>
      )}
      <Elem name="position" ref={seekerRef} style={{ left: `${seekerOffset}%` }}/>
      <Elem name="minimap">{minimap}</Elem>
    </Block>
  );
};
