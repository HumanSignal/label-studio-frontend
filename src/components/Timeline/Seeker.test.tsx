import React from 'react';
import { render } from '@testing-library/react';
import { Seeker } from './Seeker';

describe('Seeker', () => {
  it('should render', () => {
    const { container } = render(<Seeker length={1009} step={16} position={1} seekOffset={0} seekVisible={72} onIndicatorMove={() => {}} onSeek={() => {}} />);

    expect(container).toMatchSnapshot();
  });

  // it('updates indicator based on position', () => {
  //   const { } = render(<Seeker length={1009} step={16} position={1} seekOffset={0} seekVisible={72} onIndicatorMove={() => {}} onSeek={() => {}} />);

  //   act

  //   expect(container).toMatchSnapshot();
  // });
});
