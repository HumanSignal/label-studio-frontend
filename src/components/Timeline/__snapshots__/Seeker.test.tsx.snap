// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seeker should render 1`] = `
<div>
  <div
    class="dm-seeker"
  >
    <div
      class="dm-seeker__track"
    />
    <div
      class="dm-seeker__indicator"
      style="left: 0%; width: 6.206640237859267%;"
    />
    <div
      class="dm-seeker__position"
      style="left: 0.09910802775024777%;"
    />
    <div
      class="dm-seeker__minimap"
    />
  </div>
</div>
`;
